### PR TITLE
Staking rework

### PR DESCRIPTION
**Pull request by xucito**

### CODE REVIEW

File: contracts/staking/src/operations.rs [hunk 1]
```diff
@@ -644,6 +415,135 @@ pub fn create_send_msg(
...
+// Takes the earliest of the two dates
+pub fn last_time_reward_applicable(
+    current_timestamp: Uint128,
+    valid_to: Uint128,
+) -> StdResult<Uint128> {
+    if current_timestamp < valid_to {
+        return Ok(current_timestamp);
+    }
+    Ok(valid_to)
+}
```
Comment (FloppyDisck) [inline contracts/staking/src/operations.rs:427]:
```suggestion
pub fn last_time_reward_applicable(
    current_timestamp: Uint128,
    valid_to: Uint128,
) -> Uint128 {
    max(current_timestamp, valid_to)
}
```

**Review by FloppyDisck — COMMENTED** [review]
Would it be possible to separate the staking math from the contract? That way the actual staking implementation is easier to review. This would make writing tests for the math easier too.

**Review by FloppyDisck — APPROVED** [review]

**Review by DrPresident — APPROVED** [review]
Rewards logic looks good, the data in the spreadsheet also looks to be accurate so as long as the tests reflect that data it should be good to go!

File: contracts/staking/tests/integration.rs [hunk 2]
```diff
@@ -70,336 +83,601 @@ pub fn staking_integration_tests_without_proxy() {
...
+        &staking_contract,
+        OWNER_PUB_KEY,
+        OWNER_PUB_KEY,
+        get_current_block_time(&router),
     )
-    .unwrap(); 
-    
+    .unwrap();
+    assert_eq!(balances.0, 1);
+    assert_eq!(balances.1, Uint128::new(15000));
```
Comment (GumboLimbo) [inline contracts/staking/tests/integration.rs:198]: Can you help me understand why the output is 15000? We are advancing 5000 seconds, there are 86400 seconds in a day, and the daily reward rate is 300000. Shouldn't the output be (5000 / 86400) * 300000 = 17361.11?

**Comment by DrPresident** [comment]
Could we change the claimable function to reference `env.block.time` instead of a passed in timestamp?

**Review by DrPresident — CHANGES_REQUESTED** [review]
It seems the rewards rate is getting precision truncated during division, can we use the `x * 10^18` strategy to maintain that precision until we get to finalized token amounts? I believe this is causing slight yet significant truncation in rewards calculation (see alex's comment above)

**Comment by xucito** [comment]
Have updated the logic now to use reward rate * 10^18. Also added claiming rewards if you unstake

File: contracts/staking/tests/integration.rs (hunk 2 above)
Comment (xucito) [reply contracts/staking/tests/integration.rs]: This was because of precision loss, has now been resolved

**Comment by xucito** [comment]
> Could we change the claimable function to reference `env.block.time` instead of a passed in timestamp?

We could or we can pass contract_address instead of passing env. Prefer the latter as it makes it easier to test

File: contracts/staking/src/operations.rs (hunk 1 above)
Comment (xucito) [reply contracts/staking/src/operations.rs:436]: Should this not be min?

**Comment by DrPresident** [comment]
> > Could we change the claimable function to reference `env.block.time` instead of a passed in timestamp?
> 
> We could or we can pass contract_address instead of passing env. Prefer the latter as it makes it easier to test

What contract address would you pass in? I think in multi_test you can set manually set the timestamp for a block so you can just set the seconds to exactly what you want. I'll try to find an example I believe I have done it before.

**Comment by xucito** [comment]
> > > Could we change the claimable function to reference `env.block.time` instead of a passed in timestamp?
> > 
> > 
> > We could or we can pass contract_address instead of passing env. Prefer the latter as it makes it easier to test
> 
> What contract address would you pass in? I think in multi_test you can set manually set the timestamp for a block so you can just set the seconds to exactly what you want. Example: https://github.com/securesecrets/shade/blob/a797ca47e23b549fb3a47fd506322587e90ab062/contracts/dao/treasury/tests/integration/allowance.rs#L32

We use .env to just get the contract address, either way its fine

**Review by DrPresident — APPROVED** [review]

**Review by GumboLimbo — APPROVED** [review]

**Pull request merged**